File: src/ves_kinematic_calibration/ves_kinematic_calibration/palpation_kuka.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
import csv
import numpy as np
import sys

class PalpationKuka(Node):
    def __init__(self):
        super().__init__('palpation_kuka')

        # --- CONFIGURATION ---
        self.use_relative_ndi = False 
        self.push_depth = 0.006       # 6mm push
        self.retract_lift = 0.005     # 5mm lift above surface
        self.push_direction = np.array([0.0, 0.0, 1.0]) 

        # --- TOPICS ---
        self.kuka_pub = self.create_publisher(PoseStamped, '/ves/kuka/servo_cp', 10)
        self.ves_pub = self.create_publisher(PoseStamped, '/ves/left/joint/servo_cp', 10)
        
        self.ndi_sub = self.create_subscription(PoseStamped, '/sensor_pose_raw', self.ndi_callback, 10)
        self.kuka_sub = self.create_subscription(PoseStamped, '/ves/kuka/measured_cp', self.kuka_callback, 10)

        # --- STATE MACHINE ---
        self.state = "MOVE_VES_INIT" 
        self.point_index = 0
        self.recording = False
        self.current_label = ""
        
        # --- DATA BUFFERS ---
        self.last_measured_kuka = None
        self.kuka_home_pos = None
        self.kuka_home_ori = None
        self.last_cmd_pose = None
        self.ndi_ref = None
        self.ref_samples = []
        self.kuka_wait_notified = False 
        
        # Grid points (Relative to locked KUKA origin)
        self.grid_points = [
            {'x': -0.008, 'y': 0.000, 'z': 0.008},
            {'x': -0.005, 'y': -0.00, 'z': 0.008},
            {'x': -0.002, 'y': -0.00, 'z': 0.008},
            {'x': 0.001, 'y': 0.000, 'z': 0.008},
            {'x': 0.005, 'y': -0.00, 'z': 0.008},
            {'x': 0.008, 'y': -0.00, 'z': 0.008}
        ]

        # CSV Logging Setup
        self.logfile = open('kuka_palpation_python.csv', 'w', newline='')
        self.csv_writer = csv.writer(self.logfile)
        self.csv_writer.writerow(['time', 'point_index', 'phase', 'cmd_px', 'cmd_py', 'cmd_pz', 'ndi_px', 'ndi_py', 'ndi_pz'])

        # Timer (2.0s per step)
        self.timer = self.create_timer(2.0, self.state_machine)
        self.get_logger().info("🚀 Node Started. Moving VES to (0,0,0) first...")

    def kuka_callback(self, msg):
        self.last_measured_kuka = msg.pose

    def ndi_callback(self, msg):
        pos = msg.pose.position
        
        # Phase 1: Wait for VES to settle, then verify NDI signal and KUKA position
        if self.state == "WAITING_FOR_NDI":
            count = len(self.ref_samples)
            if count < 10:
                self.ref_samples.append([pos.x, pos.y, pos.z])
                if count % 2 == 0:
                    self.get_logger().info(f"📡 Collecting NDI samples: {count+1}/10")
                return

            # Once 10 NDI samples exist, check if KUKA measurement has arrived
            if self.last_measured_kuka:
                self.ndi_ref = np.mean(self.ref_samples, axis=0)
                self.kuka_home_pos = self.last_measured_kuka.position
                self.kuka_home_ori = self.last_measured_kuka.orientation
                
                self.get_logger().info("------------------------------------------")
                self.get_logger().info("✅ INITIALIZATION COMPLETE")
                self.get_logger().info(f"📍 NDI Baseline: {self.ndi_ref}")
                self.get_logger().info(f"🤖 KUKA Origin Locked.")
                self.get_logger().info("------------------------------------------")
                
                self.state = "COLLECT_BASELINE"
            else:
                if not self.kuka_wait_notified:
                    self.get_logger().warn("⏳ NDI Ready. Waiting for KUKA topic /ves/kuka/measured_cp...")
                    self.kuka_wait_notified = True
            return

        if not self.recording: return

        # Phase 2: Logging during grid execution
        t = self.get_clock().now().nanoseconds * 1e-9
        log_x = (pos.x - self.ndi_ref[0]) if self.use_relative_ndi else pos.x
        log_y = (pos.y - self.ndi_ref[1]) if self.use_relative_ndi else pos.y
        log_z = (pos.z - self.ndi_ref[2]) if self.use_relative_ndi else pos.z

        self.csv_writer.writerow([
            t, self.point_index, self.current_label,
            self.last_cmd_pose.position.x, self.last_cmd_pose.position.y, self.last_cmd_pose.position.z,
            log_x, log_y, log_z
        ])

    def get_target_msg(self, dx, dy, dz):
        msg = PoseStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = "ves/kuka/base"
        msg.pose.position.x = self.kuka_home_pos.x + dx
        msg.pose.position.y = self.kuka_home_pos.y + dy
        msg.pose.position.z = self.kuka_home_pos.z + dz
        msg.pose.orientation = self.kuka_home_ori
        self.last_cmd_pose = msg.pose
        return msg

    def state_machine(self):
        if self.state == "MOVE_VES_INIT":
            # Exact command mapping from the topic pub request
            ves_msg = PoseStamped()
            ves_msg.header.frame_id = 'world'
            ves_msg.header.stamp = self.get_clock().now().to_msg()
            ves_msg.pose.position.x = 0.0 
            ves_msg.pose.position.y = 0.0 
            ves_msg.pose.position.z = 0.0 
            ves_msg.pose.orientation.x = 0.0
            ves_msg.pose.orientation.y = 0.0
            ves_msg.pose.orientation.z = 0.0
            ves_msg.pose.orientation.w = 1.0
            self.ves_pub.publish(ves_msg)
            
            self.get_logger().info("📦 VES positioned to (0,0,0). Settling for 3s...")
            self.ves_settle_start = self.get_clock().now()
            self.state = "VES_SETTLING"

        elif self.state == "VES_SETTLING":
            # Wait for mechanical vibrations to stop
            elapsed = (self.get_clock().now() - self.ves_settle_start).nanoseconds / 1e9
            if elapsed >= 3.0:
                self.get_logger().info("✅ VES Stable. Collecting NDI/KUKA origins...")
                self.state = "WAITING_FOR_NDI"

        elif self.state == "WAITING_FOR_NDI":
            return

        elif self.state == "COLLECT_BASELINE":
            self.start_record("baseline", True)
            self.state = "MOVE_TO_SURFACE"

        elif self.state == "MOVE_TO_SURFACE":
            self.recording = False
            pt = self.grid_points[self.point_index]
            self.kuka_pub.publish(self.get_target_msg(pt['x'], pt['y'], pt['z']))
            self.state = "PAUSE_SURFACE"

        elif self.state == "PAUSE_SURFACE":
            self.start_record("surface", True)
            self.state = "MOVE_TO_PUSH"

        elif self.state == "MOVE_TO_PUSH":
            self.recording = False
            pt = self.grid_points[self.point_index]
            px = pt['x'] + (self.push_depth * self.push_direction[0])
            py = pt['y'] + (self.push_depth * self.push_direction[1])
            pz = pt['z'] + (self.push_depth * self.push_direction[2])
            self.kuka_pub.publish(self.get_target_msg(px, py, pz))
            self.state = "PAUSE_PUSH"

        elif self.state == "PAUSE_PUSH":
            self.start_record("push", True)
            self.state = "RETRACT"

        elif self.state == "RETRACT":
            self.recording = False
            pt = self.grid_points[self.point_index]
            # Safety lift 5mm above surface
            self.kuka_pub.publish(self.get_target_msg(pt['x'], pt['y'], pt['z'] - self.retract_lift))
            self.state = "ADVANCE_POINT"

        elif self.state == "ADVANCE_POINT":
            self.point_index += 1
            if self.point_index >= len(self.grid_points):
                self.state = "DONE"
                self.get_logger().info("✅ Sequence Complete.")
            else:
                self.state = "MOVE_TO_SURFACE"

    def start_record(self, label, rec):
        self.current_label = label
        self.recording = rec
        self.get_logger().info(f"▶️ Phase: {label} | Point: {self.point_index}")

def main():
    rclpy.init()
    node = PalpationKuka()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.logfile.close()
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/ves_kinematic_calibration/ves_kinematic_calibration/palpation_kuka.cpp
#include <fstream>
#include <vector>
#include <chrono>
#include <string>
#include <iomanip>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"

using namespace std::chrono_literals;

enum State {
    WAITING_FOR_NDI_SIGNAL, // Mandatorily collect 10 samples to verify signal
    COLLECT_BASELINE,       // Phase to record initial baseline for the log
    MOVE_TO_SURFACE,        // Navigate to grid point
    PAUSE_SURFACE,          // 2s pause at surface
    MOVE_TO_PUSH,           // Execute push
    PAUSE_PUSH,             // 2s pause at depth
    RETRACT,                // Lift slightly above point
    ADVANCE_POINT,          // Move to next or finish
    DONE
};

struct Point3D { double x, y, z; };

class KukaPalpator : public rclcpp::Node {
public:
    KukaPalpator() : Node("kuka_palpator"), state_(WAITING_FOR_NDI_SIGNAL), point_index_(0), recording_(false) {

        // --- CONFIG FLAG ---
        // Set to TRUE for (Current - Baseline), FALSE for Raw Absolute coordinates
        use_relative_ndi_ = true; 

        kuka_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("/ves/kuka/servo_cp", 10);
        ves_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("/ves/left/joint/servo_cp", 10);

        ndi_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "/sensor_pose_raw", 10, std::bind(&KukaPalpator::ndi_callback, this, std::placeholders::_1));

        kuka_measured_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
            "/ves/kuka/measured_cp", 10, [this](const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
                last_measured_kuka_ = msg->pose;
            });

        logfile_.open("kuka_palpation_experiment.csv");
        // Column headers
        logfile_ << "time,point_index,phase,cmd_px,cmd_py,cmd_pz,ndi_px,ndi_py,ndi_pz\n";

        // Define your palpation grid points (X, Y, Z relative to start)
        grid_points_ = {
            {-0.008, 0.000, 0.008}, {-0.005, -0.001, 0.008}, {-0.002, -0.006, 0.008},
            {-0.008, 0.000, 0.012}, {-0.005, -0.002, 0.012}, {-0.001, -0.004, 0.012}
        };

        push_depth_ = 0.006;
        retract_lift_ = 0.005; 
        push_direction_ = {0.0, 0.0, 1.0}; // Direction vector for push action

        timer_ = this->create_wall_timer(2000ms, std::bind(&KukaPalpator::state_machine, this));
        
        RCLCPP_INFO(this->get_logger(), "Node initialized. Mode: %s. Waiting for 10 NDI samples...", 
                    use_relative_ndi_ ? "RELATIVE" : "RAW");
    }

private:
    geometry_msgs::msg::PoseStamped get_target_msg(double dx, double dy, double dz) {
        geometry_msgs::msg::PoseStamped msg;
        msg.header.stamp = this->get_clock()->now();
        msg.header.frame_id = "ves/kuka/base";
        msg.pose.position.x = kuka_home_pos_.x + dx;
        msg.pose.position.y = kuka_home_pos_.y + dy;
        msg.pose.position.z = kuka_home_pos_.z + dz;
        msg.pose.orientation = kuka_home_ori_;
        last_cmd_pose_ = msg.pose;
        return msg;
    }

    void ndi_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
        // ALWAYS collect 10 samples first to verify NDI is alive and calculate baseline
        if (state_ == WAITING_FOR_NDI_SIGNAL) {
            ref_samples_.push_back(msg->pose.position);
            if (ref_samples_.size() >= 10) {
                double sx=0, sy=0, sz=0;
                for(auto& p : ref_samples_) { sx+=p.x; sy+=p.y; sz+=p.z; }
                ndi_ref_ = {sx/10.0, sy/10.0, sz/10.0};
                
                // Lock current KUKA position as the origin for the relative grid
                kuka_home_pos_ = last_measured_kuka_.position;
                kuka_home_ori_ = last_measured_kuka_.orientation;
                
                RCLCPP_INFO(this->get_logger(), "✅ NDI Signal Verified. Baseline Calculated. Starting Baseline Phase.");
                state_ = COLLECT_BASELINE;
            }
            return;
        }

        if (!recording_) return;

        double t = this->get_clock()->now().nanoseconds() * 1e-9;
        
        // Logical Switch for Logging: uses the baseline calculated during the 10-sample wait
        double log_x = use_relative_ndi_ ? (msg->pose.position.x - ndi_ref_.x) : msg->pose.position.x;
        double log_y = use_relative_ndi_ ? (msg->pose.position.y - ndi_ref_.y) : msg->pose.position.y;
        double log_z = use_relative_ndi_ ? (msg->pose.position.z - ndi_ref_.z) : msg->pose.position.z;

        logfile_ << std::fixed << std::setprecision(6) << t << "," << point_index_ << "," << current_label_ << ","
                 << last_cmd_pose_.position.x << "," << last_cmd_pose_.position.y << "," << last_cmd_pose_.position.z << ","
                 << log_x << "," << log_y << "," << log_z << "\n";
    }

    void state_machine() {
        switch (state_) {
            case COLLECT_BASELINE:
                start_record("baseline", true);
                state_ = MOVE_TO_SURFACE;
                break;

            case MOVE_TO_SURFACE:
                recording_ = false;
                kuka_pub_->publish(get_target_msg(grid_points_[point_index_].x, grid_points_[point_index_].y, grid_points_[point_index_].z));
                state_ = PAUSE_SURFACE;
                break;

            case PAUSE_SURFACE:
                start_record("surface", true);
                state_ = MOVE_TO_PUSH;
                break;

            case MOVE_TO_PUSH: {
                recording_ = false;
                double px = grid_points_[point_index_].x + (push_depth_ * push_direction_.x);
                double py = grid_points_[point_index_].y + (push_depth_ * push_direction_.y);
                double pz = grid_points_[point_index_].z + (push_depth_ * push_direction_.z);
                kuka_pub_->publish(get_target_msg(px, py, pz));
                state_ = PAUSE_PUSH;
                break;
            }

            case PAUSE_PUSH:
                start_record("push", true);
                state_ = RETRACT;
                break;

            case RETRACT:
                recording_ = false;
                // Lift safely above surface (e.g. 5mm lift)
                kuka_pub_->publish(get_target_msg(grid_points_[point_index_].x, grid_points_[point_index_].y, grid_points_[point_index_].z - retract_lift_));
                state_ = ADVANCE_POINT;
                break;

            case ADVANCE_POINT:
                point_index_++;
                state_ = (point_index_ >= grid_points_.size()) ? DONE : MOVE_TO_SURFACE;
                break;

            case DONE:
                RCLCPP_INFO(this->get_logger(), "✅ All grid points completed successfully.");
                break;

            default:
                break;
        }
    }

    void start_record(std::string label, bool rec) {
        current_label_ = label;
        recording_ = rec;
        RCLCPP_INFO(this->get_logger(), "Phase: %s | Current Point Index: %zu", label.c_str(), point_index_);
    }

    // ROS 2 Comms
    rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr kuka_pub_, ves_pub_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr ndi_sub_, kuka_measured_sub_;
    rclcpp::TimerBase::SharedPtr timer_;

    // Data handling
    std::ofstream logfile_;
    State state_;
    size_t point_index_;
    bool recording_, use_relative_ndi_;
    std::string current_label_;

    // Hardware and Sensor Buffers
    geometry_msgs::msg::Pose last_measured_kuka_, last_cmd_pose_;
    geometry_msgs::msg::Point kuka_home_pos_, ndi_ref_;
    geometry_msgs::msg::Quaternion kuka_home_ori_;
    std::vector<geometry_msgs::msg::Point> ref_samples_;
    
    // Grid Specs
    std::vector<Point3D> grid_points_;
    double push_depth_, retract_lift_;
    Point3D push_direction_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<KukaPalpator>());
    rclcpp::shutdown();
    return 0;
}